gc.c

```diff
@@ -7208,6 +7208,47 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
     }
 }
 
+static bool
+gc_declarative_marking_p(const rb_data_type_t *type)
+{
+    return (type->flags & RUBY_TYPED_DECL_MARKING) != 0;
+}
+
+static inline void
+gc_mark_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
+{
+    VALUE markable = *(data_struct + (offset / sizeof(VALUE)));
+    rb_gc_mark_movable(markable);
+}
+
+static inline void
+gc_ref_update_from_offset(rb_objspace_t *objspace, VALUE *data_struct, VALUE offset)
```

Comment:
- Use `size_t` for offsets

```suggestion
gc_ref_update_from_offset(rb_objspace_t *objspace, void *data_struct, size_t offset)
```